### ISSUE TITLE
Exception is thrown during parsing of response to a request which triggered Azure's content management

### ISSUE THREAD

**Issue by wskinner-ripcord** [openai/openai-python#349]
### Describe the bug

When the Azure content management system flags a request, the library fails to handle the response, causing an exception to be thrown.

Example stack trace:
```
Traceback (most recent call last):
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_requestor.py", line 331, in handle_error_response
    error_data = resp["error"]
TypeError: string indices must be integers

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/WSkinner/work/ripcord/ml-generative/bug.py", line 9, in <module>
    response = openai.Completion.create(
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_resources/completion.py", line 25, in create
    return super().create(*args, **kwargs)
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 153, in create
    response, _, api_key = requestor.request(
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_requestor.py", line 226, in request
    resp, got_stream = self._interpret_response(result, stream)
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_requestor.py", line 619, in _interpret_response
    self._interpret_response_line(
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_requestor.py", line 682, in _interpret_response_line
    raise self.handle_error_response(
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/api_requestor.py", line 333, in handle_error_response
    raise error.APIError(
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/error.py", line 32, in __init__
    self.error = self.construct_error_object()
  File "/Users/WSkinner/.pyenv/versions/3.10.7/envs/ml-generative/lib/python3.10/site-packages/openai/error.py", line 62, in construct_error_object
    or not isinstance(self.json_body["error"], dict)
TypeError: string indices must be integers
```

There is a [related issue](https://github.com/openai/openai-python/issues/331), from which I have copied the prompt which triggers the content management policy. However, that issue appears to be treating the rejection of a request due to content management, resulting in a 400 status code, as a bug. This new issue is specifically regarding the `openai-python` library's treatment of that response, not the fact that the response was returned.

### To Reproduce

To reproduce the issue, run the following code.
```python
import os
import openai

openai.api_type = "azure"
openai.api_version = "2023-03-15-preview"
openai.api_base = "https://rc-ai.openai.azure.com/"
openai.api_key = os.getenv("AZURE_OPENAI_API_KEY")

response = openai.Completion.create(
    engine="gpt-35-turbo",
    prompt="SUBREDDIT: r/AskReddit TITLE: Cock blocked by a friend (Who's a girl). POST: So for the past week there's "
        "been this girl in one of my classes I've been talking to, she's pretty cute (dyed red hair, fair skin, "
        "a few freckles, not ginger), she loves star wars and I suspect she's a redditor. I was going to ask her for "
        "her number today, but a girl i met about a year ago came and sat right where the red head had been sitting, "
        "effectively cock-blocking me and driving the girl I was interested in away. Now it seems like the red head "
        "thinks I'm uninterested in her and has since found some other guy to talk to. Has anybody been in a similar "
        "scenario? Advice? \nTL;DR: Got cock blocked by a friend who's a girl."
)
```

### OS

macOS

### Python version

3.10.7

### Library version

0.27.2

**Comment by hallacy** [comment]
Ah, now #331 makes sense.

I don't have access to the Azure endpoints.  If you run a curl command instead of using this python repo, what response do you get? That should help me figure out where our parsing went wrong

**Comment by wskinner-ripcord** [comment]
@hallacy 

```bash
curl $OPENAI_API_BASE/openai/deployments/gpt-35-turbo/chat/completions\?api-version\=2023-03-15-preview \
  -H "Content-Type: application/json" \
  -H "api-key: $OPENAI_API_KEY" \
  -d '{"messages": [{"role": "user", "content": "SUBREDDIT: r/AskReddit TITLE: Cock blocked by a friend (Whos a girl. POST: So for the past week theres been this girl in one of my classes Ive been talking to, shes pretty cute (dyed red hair, fair skin, a few freckles, not ginger, she loves star wars and I suspect shes a redditor. I was going to ask her for her number today, but a girl i met about a year ago came and sat right where the red head had been sitting, effectively cock-blocking me and driving the girl I was interested in away. Now it seems like the red head thinks Im uninterested in her and has since found some other guy to talk to. Has anybody been in a similar scenario? Advice? TL;DR: Got cock blocked by a friend whos a girl."}]}'

{"error":{"message":"The response was filtered due to the prompt triggering Azure OpenAI’s content management policy. Please modify your prompt and retry. To learn more about our content filtering policies please read our documentation: https://go.microsoft.com/fwlink/?linkid=2198766","type":null,"param":"prompt","code":"content_filter","status":400}}
```

**Comment by StevenTang1998** [comment]
The same bug. Is there any solution?

**Comment by wskinner-ripcord** [comment]
@StevenTang1998 Azure content filtering seems to be pretty naive. I was able to get decent results by using a list of stopwords and removing all matching words from my prompts. This is a crude solution and may or may not work for your use case. Unless you have a significant partnership with Azure, they won't remove the filtering. The other solution is to use the OpenAI API. In my experience OpenAI has slower response times and lower rate limits, but they don't do any content filtering of prompts.

**Comment by wskinner-ripcord** [comment]
@StevenTang1998 You can wrap the request in a `try... except` block and write some logic to handle the error. Or use OpenAI instead of Azure.

**Comment by lorr1** [comment]
Looks like `self.json_body` is a string. Need to `json.loads(self.json_body)` somewhere.

**Comment by hallacy** [comment]
Huh.  @cmurtz-msft @sorinsuciu-msft I dont suppose one of you would know why this would be happening? At a first glance, this should only occur if the `Content-Type` being returned is text/plain

**Comment by hallacy** [comment]
Related: https://github.com/openai/openai-python/commit/931b4d23c742349bf5f82a2a995dd46ae684cdd5 was merged to  at least prevent this from throwing an exception.

**Comment by ilia10000** [comment]
Hey folks! Any updates on this bug?

**Comment by kristapratico** [comment]
hey @ilia10000, I want to make sure I understand your question since there are a few things going on in this issue.

The original bug opened described how the openai library was throwing `TypeError: string indices must be integers` because the Azure OpenAI service is returning an unexpected content-type of `text/plain` when the content filter is triggered on an inappropriate prompt. A [PR](https://github.com/openai/openai-python/commit/931b4d23c742349bf5f82a2a995dd46ae684cdd5) was merged to handle this more gracefully, i.e. raise an APIError with the error message:

> openai.error.APIError: Invalid response object from API: '{"error":{"message":"The response was filtered due to the prompt triggering Azure OpenAI’s content management policy. Please modify your prompt and retry. To learn more about our content filtering policies please read our documentation: https://go.microsoft.com/fwlink/?linkid=2198766","type":null,"param":"prompt","code":"content_filter","status":400}}' (HTTP response code was 400)

This fix was merged and released in versions 0.27.3 or greater. 

The cause of the issue (returning text/plain instead of application/json) is still a problem and I will follow-up on this issue with the Azure service team.

Is your question about the above or is it about the content filter raising an exception on the prompt?

**Comment by kristapratico** [comment]
@JensMadsen that seems reasonable to me. Let me bring it up with the team.

**Comment by JensMadsen** [comment]
@kristapratico Thank you soo much for a very swift reply and action!!

**Comment by JensMadsen** [comment]
@kristapratico I still get `openai.error.APIError` with `openai` v0.27.7. I use a privately deployed Azure model (in EU) though. I have also tried to change: ` openai.api_version = '2022-12-01'`  -> `openai.api_version ='2023-03-15-preview'`, but still no cigar?

**Comment by kristapratico** [comment]
@JensMadsen can you provide your endpoint name and the region (West Europe, UK West, etc.) where the model is deployed? If you would rather email it to me, you can reach me at krpratic@microsoft.com. It's possible that the model will need to be redeployed since the issue was with the service underneath, not the API or SDK.

**Comment by solyarisoftware** [comment]
See also: https://github.com/openai/openai-python/issues/331#issuecomment-1784833243

**Comment by rattrayalex** [comment]
The bug in the Python library that was reported here has been fixed in v1.
